table/mem_table.go

```diff
@@ -0,0 +1,231 @@
+
+func (mt *MemTable) DecrRef() {
+	mt.skl.DecrRef()
+}
+
+func (mt *MemTable) Empty() bool {
+	return atomic.LoadPointer(&mt.pendingList) == nil && mt.skl.Empty()
+}
+
+func (mt *MemTable) Get(key []byte) y.ValueStruct {
+	curr := (*listNode)(atomic.LoadPointer(&mt.pendingList))
+	for curr != nil {
+		if curr.isMerged() {
+			break
+		}
+		if v, ok := curr.get(key); ok {
+			return v
+		}
+		curr = (*listNode)(curr.next)
+	}
+	return mt.skl.Get(key)
+}
+
+func (mt *MemTable) NewIterator(reverse bool) y.Iterator {
+	var (
+		sklItr = mt.skl.NewUniIterator(reverse)
+		its    []y.Iterator
+	)
+	curr := (*listNode)(atomic.LoadPointer(&mt.pendingList))
+	for curr != nil {
+		if curr.isMerged() {
+			break
+		}
+		its = append(its, curr.newIterator(reverse))
+		curr = (*listNode)(curr.next)
+	}
+
+	if len(its) == 0 {
+		return sklItr
+	}
+	its = append(its, sklItr)
+	return NewMergeIterator(its, reverse)
+}
+
+func (mt *MemTable) MemSize() int64 {
+	var sz int64
+	curr := (*listNode)(atomic.LoadPointer(&mt.pendingList))
+	for curr != nil {
+		sz += curr.memSize
+		curr = (*listNode)(curr.next)
+	}
+	return mt.skl.MemSize() + sz
+}
+
+// PutToSkl directly insert entry into SkipList.
+func (mt *MemTable) PutToSkl(key []byte, v y.ValueStruct) {
+	mt.skl.Put(key, v)
+}
+
+// PutToPendingList put entries to pending list, and you can call MergeListToSkl to merge them to SkipList later.
+func (mt *MemTable) PutToPendingList(entries []Entry) {
+	mt.putToList(entries)
+}
+
+// MergeListToSkl merge all entries in pending list to SkipList.
+func (mt *MemTable) MergeListToSkl() {
+	head := (*listNode)(atomic.LoadPointer(&mt.pendingList))
+	if head == nil {
+		return
+	}
+
+	head = (*listNode)(atomic.LoadPointer(&mt.pendingList))
+	if head == nil {
+		return
+	}
+	mt.mergeListToSkl(head)
+
+	// No new node inserted, just update head of list.
+	if atomic.CompareAndSwapPointer(&mt.pendingList, unsafe.Pointer(head), nil) {
+		return
+	}
+	// New node inserted, iterate to find `prev` of old head.
+	curr := (*listNode)(atomic.LoadPointer(&mt.pendingList))
+	for curr != nil {
+		next := atomic.LoadPointer(&curr.next)
+		if unsafe.Pointer(head) == next {
+			atomic.StorePointer(&curr.next, nil)
+			return
+		}
+		curr = (*listNode)(next)
+	}
+}
+
+func (mt *MemTable) putToSkl(entries []Entry) {
+	var hint skl.Hint
+	for _, e := range entries {
+		mt.skl.PutWithHint(e.Key, e.Value, &hint)
+	}
+}
+
+func (mt *MemTable) putToList(entries []Entry) {
+	n := newListNode(entries)
+	for {
+		old := atomic.LoadPointer(&mt.pendingList)
+		n.next = old
+		if atomic.CompareAndSwapPointer(&mt.pendingList, old, unsafe.Pointer(n)) {
+			return
+		}
+	}
+}
+
+func (mt *MemTable) mergeListToSkl(n *listNode) {
+	next := (*listNode)(atomic.LoadPointer(&n.next))
+	if next != nil {
+		mt.mergeListToSkl(next)
+	}
+	atomic.StorePointer(&n.next, nil)
+	mt.putToSkl(n.entries)
+	atomic.StoreUint32(&n.merged, 1)
+}
+
+type listNode struct {
+	next    unsafe.Pointer // *listNode
+	merged  uint32
```

Comment:
Can we remove this field `merged` to make it simpler?
We can still use this node even if it is merged.

Reply:
Just considering that if `Get` or `Iterate` starts at the same time as `MergeList` is completed, having `merged` can reduce unnecessary overhead, because it can ignore `pendingList` ASAP.

Reply:
We can still check `head` of memtable and `next` of the node to determine if the node is merged.

Reply:
The last node doesn't have `next`😂
I'll remove `merged`